Fix the input tools format and history tool_calls in OpenAI API

chen700564 opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->
Adjust the input for the OpenAI API

1. Keep the full `tool` dictionary in the input, consistent with vLLM, instead of only retaining the `function` field.

2. After a `tool_call` is generated, it is structured as:
```python
{
  ...,
  "function": {
    "name": "xxx",
    "arguments": "json.dumps(...)"
  }
}
```

When `tool_call` appears in the conversation history, using `tokenizer.apply_chat_template` causes the `arguments` part to be the result of `json.dumps`, which does not align with the original output. Therefore, during conversation history processing, the `arguments` should be parsed using `json.loads`.


## Modifications

<!-- Describe the changes made in this PR. -->

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

chen700564 committed:
keep full tools in request

chen700564 committed:
fix the type of function.arguments in message

chen700564 committed:
Merge branch 'sgl-project:main' into tool_call_fix

chen700564 committed:
Merge branch 'sgl-project:main' into tool_call_fix

chen700564 committed:
update

zhaochenyang20 committed:
Merge branch 'main' into tool_call_fix

shuaills commented:
CC @CatherineSue

CatherineSue reviewed (commented):
Overall LGTM. Can we apply all gemini's comments? Those all valid points

chen700564 committed:
Merge branch 'main' into tool_call_fix

chen700564 committed:
Update adapter.py

chen700564 committed:
Update adapter.py

chen700564 commented:
I have applied gemini's comments.

chen700564 committed:
Merge branch 'main' into tool_call_fix

CatherineSue commented on python/sglang/srt/openai_api/adapter.py:
```diff
@@ -1082,18 +1082,44 @@ def v1_chat_generate_request(
+                                logger.warning(f"Failed to parse tool call arguments as JSON: {e}")
+                                # Decide whether to continue or raise the exception based on desired behavior
+                                continue # Or raise e if strict parsing is required
+                    if isinstance(processed_msg.get("content"), list):
```
What is this block for?

chen700564 replied:
This block is used to handle and validate the arguments in a tool call when using the qwen25 tool parser, ensuring that the arguments can be successfully parsed by `json.loads()`. 

 When the model generates a tool call, it should also output valid JSON-formatted arguments. However, in some cases, the generated arguments may not be valid JSON (e.g., due to syntax errors or incorrect formatting), which would cause `json.loads()` to throw an exception when attempting to parse them.

CatherineSue commented on python/sglang/srt/openai_api/adapter.py:
```diff
@@ -1082,18 +1082,44 @@ def v1_chat_generate_request(
+                                openai_compatible_messages.append(new_msg)
+                    else:
+                        openai_compatible_messages.append(processed_msg)
                 if (
```
Please add back the comment at L1096 originally here

CatherineSue commented:
Can we update the PR title as well?

chen700564 committed:
Add back the comment at L1096 originally

chen700564 committed:
remove out-of-date code block

CatherineSue commented:
@chen700564 Can you run a pre-commit, I think there are lint errors. And Can you update the PR title to include a bit more information? Thanks!

chen700564 committed:
Merge branch 'sgl-project:main' into tool_call_fix

chen700564 committed:
pre-commit fix

chen700564 committed:
Delete python/sglang/srt/openai_api/adapter.py

chen700564 committed:
Merge branch 'sgl-project:main' into tool_call_fix

chen700564 committed:
new version: keep full tools in request

chen700564 committed:
new version: fix the type of function.arguments in OpenAI message

chen700564 committed:
pre-commit fix

chen700564 committed:
Merge branch 'main' into tool_call_fix

chen700564 commented:
> > I've run the pre-commit and fixed the linting issues. Also, I've updated the PR title to provide more context. Is there anything else I need to modify?
> 
> Hi, @chen700564 , we recently refactored our openai server to `python/sglang/srt/entrypoints/openai`. Could you update your changes here? I apologize for the tedious process due to the refactor.
> 
> https://github.com/sgl-project/sglang/blob/72676cd6c0eead68286cf9c8df5c67f580ff477b/python/sglang/srt/entrypoints/openai/serving_chat.py#L130-L136
> 
> https://github.com/sgl-project/sglang/blob/72676cd6c0eead68286cf9c8df5c67f580ff477b/python/sglang/srt/entrypoints/openai/serving_chat.py#L193-L201

Hello, I have updated my changes in the new version.

CatherineSue reviewed (approved):
LGTM. Let's wait for the CI to finish

CatherineSue committed:
Merge branch 'main' into tool_call_fix

CatherineSue committed:
Merge branch 'main' into tool_call_fix

chen700564 committed:
Merge branch 'main' into tool_call_fix

chen700564 committed:
Merge branch 'main' into tool_call_fix

YosanHo commented:
@CatherineSue @zhyncs  This change is not compatible with other models, such as DeepSeek. it's not bug of sglang, Please revert this PR , k2 is updated template, https://huggingface.co/moonshotai/Kimi-K2-Instruct/commit/6a1f5e6be5b1ddfd220836faed223dd70fbc091e

CatherineSue commented:
@YosanHo thanks for reporting this. We have opened a revert PR.
